### DISCUSSION TITLE
Configure Elsa 3  migration to SQL Server

### DISCUSSION THREAD

**Discussion by leostone68** [elsa-workflows/elsa-core#5706] [Design & tech]
I need to configure Elsa to use SQL Server and perform the migration. Then tell me what packages to install for Elsa 3.1.3?

**Comment by glime-ai** [comment]
To configure Elsa to use SQL Server for persistence in Elsa 3.1.3, you will need to install the appropriate NuGet packages and configure your application accordingly. Here are the steps you should follow:

1. Install the necessary Elsa 3.x NuGet packages for Entity Framework Core and SQL Server support. You can use the following commands:

   ```
   dotnet add package Elsa
   dotnet add package Elsa.EntityFrameworkCore.SqlServer
   ```

2. In your `Program.cs` or startup configuration file, set up Elsa with Entity Framework Core using the SQL Server provider. Here's an example of how you might configure it:

   ```csharp
   builder.Services.AddElsa(elsa =>
   {
       elsa.UseEntityFrameworkPersistence(ef => ef.UseSqlServer("YourConnectionString"));
   });
   ```

   Replace `"YourConnectionString"` with your actual SQL Server connection string.

3. Perform the migration to create the necessary tables in your SQL Server database. You can use the following command:

   ```
   dotnet ef migrations add InitialCreate -o Data/Migrations
   dotnet ef database update
   ```

   Make sure that the Entity Framework Core tools are installed for these commands to work. If not, you can install them using:

   ```
   dotnet tool install --global dotnet-ef
   ```

4. Ensure that your application is correctly configured to use the Elsa services and middleware as needed.

By following these steps and using the provided commands, you should be able to set up Elsa to use SQL Server for persistence in your application with Elsa 3.1.3.

This comment was generated by [Glime](https://glimelab.ai/?utm_source=dRbvTb7IhCSYiI&utm_medium=github).

**Reply by leostone68** [reply]
not wrorking...
Resolved with: 
elsa.UseWorkflowManagement(management => management.UseEntityFrameworkCore(ef => ef.UseSqlServer(sqlServerConnectionString)));

**Comment by NickKellett** [comment]
In 3.5.0 if you configure Identify, Runtime, and Management to use EF Core, there appear to be different DbContexts for each and you have to specify these when running the dotnet EF commands in Visual Studio CLI or Powershell.  To get migrations working in a custom project using Elsa I therefore had to run these commands: 

`dotnet ef migrations add InitialCreate -o Data/Migrations/Management --context Elsa.EntityFrameworkCore.Modules.Management.ManagementElsaDbContext`
`dotnet ef migrations add InitialCreate -o Data/Migrations/Runtime --context Elsa.EntityFrameworkCore.Modules.Runtime.RuntimeElsaDbContext`
`dotnet ef migrations add InitialCreate -o Data/Migrations/Identity --context Elsa.EntityFrameworkCore.Modules.Identity.IdentityElsaDbContext`
`dotnet ef database update --context Elsa.EntityFrameworkCore.Modules.Management.ManagementElsaDbContext`
`dotnet ef database update --context Elsa.EntityFrameworkCore.Modules.Runtime.RuntimeElsaDbContext`
`dotnet ef database update --context Elsa.EntityFrameworkCore.Modules.Identity.IdentityElsaDbContext`

Luckily there is only one shared Migration Table and you probably only have to run these one time.

**Comment by sfmskywalker** [comment]
Closing this as resolved. For Elsa `3.1.3`, SQL Server persistence is configured on the individual Elsa features, for example `UseWorkflowManagement(management => management.UseEntityFrameworkCore(ef => ef.UseSqlServer(connectionString)))`, as noted in the resolved reply above.

The 3.1.3 server bundle follows the same pattern for Identity, Workflow Management, and Workflow Runtime, and sets `ef.RunMigrations` on those EF Core features. It also ships the SQL Server provider and migrations in `Elsa.EntityFrameworkCore.SqlServer`; the earlier `UseEntityFrameworkPersistence` example is not the applicable configuration API for this version.
